src/main/java/org/jabref/gui/maintable/MainTable.java

```diff
@@ -279,62 +278,75 @@ public void paste() {
             event.acceptTransferModes(TransferMode.COPY, TransferMode.LINK);
         }
+        event.consume();
     }
 
     private void handleOnDragEntered(TableRow<BibEntryTableViewModel> row, BibEntryTableViewModel entry, MouseDragEvent event) {
         // Support the following gesture to select entries: click on one row -> hold mouse button -> move over other rows
         // We need to select all items between the starting row and the row where the user currently hovers the mouse over
         // It is not enough to just select the currently hovered row since then sometimes rows are not marked selected if the user moves to fast
-        TableRow<BibEntryTableViewModel> sourceRow = (TableRow<BibEntryTableViewModel>) event.getGestureSource();
-        getSelectionModel().selectRange(sourceRow.getIndex(), row.getIndex());
+        //  TableRow<BibEntryTableViewModel> sourceRow = (TableRow<BibEntryTableViewModel>) event.getGestureSource();
+        //  getSelectionModel().selectRange(sourceRow.getIndex(), row.getIndex());
+        event.consume();
     }
 
     private void handleOnDragDetected(TableRow<BibEntryTableViewModel> row, BibEntryTableViewModel entry, MouseEvent event) {
         // Start drag'n'drop
-        row.startFullDrag();
+        //row.startFullDrag();
 
-        List<BibEntry> entries = getSelectionModel().getSelectedItems().stream().map(BibEntryTableViewModel::getEntry).collect(Collectors.toList());
+        if (!row.isEmpty()) {
+            Integer index = row.getIndex();
+            Dragboard db = row.startDragAndDrop(TransferMode.MOVE);
+            db.setDragView(row.snapshot(null, null));
+            ClipboardContent cc = new ClipboardContent();
+            cc.put(SERIALIZED_MIME_TYPE, index);
+            db.setContent(cc);
 
-        if (entries != null) {
-            ClipboardContent content = new ClipboardContent();
-            Dragboard dragboard = startDragAndDrop(TransferMode.MOVE);
-            content.put(DragAndDropDataFormats.ENTRIES, "");
-            dragboard.setContent(content);
+            List<BibEntry> entries = getSelectionModel().getSelectedItems().stream().map(BibEntryTableViewModel::getEntry).collect(Collectors.toList());
             LocalDragboard.INSTANCE.putValue(DragAndDropDataFormats.BIBENTRY_LIST_CLASS, entries);
+            event.consume();
         }
-
-        event.consume();
     }
 
-    private void handleOnDragDropped(BibEntryTableViewModel originalItem, DragEvent event) {
+    private void handleOnDragDropped(TableRow<BibEntryTableViewModel> row, BibEntryTableViewModel originalItem, DragEvent event) {
 
-        boolean success = false;
+        Dragboard db = event.getDragboard();
+        if (db.hasContent(SERIALIZED_MIME_TYPE)) {
+            int draggedIndex = (Integer) db.getContent(SERIALIZED_MIME_TYPE);
 
-        ObservableList<BibEntryTableViewModel> items = this.itemsProperty().get();
+            BibEntry entry = this.getItems().get(draggedIndex).getEntry();
```

Comment:
I followed the guide in https://stackoverflow.com/a/28606524/3450689 
@tobiasdiez  How do I implement the insert/delete at best? The problem is that all those ViewModel stuff is wrapped in many Unmodiefiable/ReadOnly Collections which don't support remove/insert 

And calling database.removeEntry/insertEntry everytime I just do drag and drop on the maintable seems to be a bit overhead (as all kind of events are triggering)
Or do we just don't allow reordering the MainTable?

Reply:
I think reordering of entries is a delicate business: what does it mean to reorder entries if I sort the table by author? The order of the entries in the bib file is determined independently anyway. For these reasons, I would not try to implement reordering of entries.